Fix query failure when referencing a field of a NULL row

geraint0923 opened the pull request:
We try to pre-calculate the row field value if possible when planning.
Before this commit, we fail to handle the case when the row itself is
evaluated to be null, which will lead to a NullPointerException.

In this commit, the row field will be pre-calculated as a null if the
row is evaluated to be a null.

electrum commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -2646,6 +2646,16 @@ public void testSimpleLeftJoin()
+        // With null base in outer side dereference expression in condition
+        assertQuery(
+                "SELECT x.val FROM " +
+                        "(SELECT cast(row(v) as row(val integer)) FROM (values 1,2,3) t(v)) ta(x) " +
```
Capitalize `VALUES` and add spaces after comma

Add space between `ta` and `(`

electrum commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -2646,6 +2646,16 @@ public void testSimpleLeftJoin()
+                        "(SELECT cast(row(v) as row(val integer)) FROM (values 1,2,3) t(v)) ta(x) " +
+                        "LEFT OUTER JOIN " +
+                        "(SELECT cast(row(v) as row(val integer)) FROM (values 1,2,3) t(v)) tb(y) " +
+                        "ON x.val=y.val " +
```
Space around `=`

electrum commented:
This commit message doesn't make sense. Explain in terms the user can understand. This is a fix to the expression interpreter, so this seems to only happen during planning? And for a join?

electrum commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -2646,6 +2646,16 @@ public void testSimpleLeftJoin()
         assertQuery("SELECT COUNT(*) FROM lineitem LEFT JOIN orders ON lineitem.orderkey = orders.orderkey");
         assertQuery("SELECT COUNT(*) FROM lineitem LEFT OUTER JOIN orders ON lineitem.orderkey = orders.orderkey");
+
+        // With null base in outer side dereference expression in condition
```
This doesn't make sense

geraint0923 commented:
@electrum Addressed the comments and updated the commit message.

I used left outer join query as the test case because it is the only query that I could use to reproduce this issue. I have considered using `cast(null as row(...)).row_field` but it seems that `cast(null as row(..))` will not be evaluated to be null in planning time.

cberner commented:
I would word the commit message more like: `Fix query failure when referencing a field of a NULL row`. For a simple fix like this, you don't need a lot of detail in the commit message

cberner commented:
Looks good as long as Travis passes
